var/spack/repos/builtin/packages/occa/package.py

```diff
@@ -0,0 +1,152 @@
+# Created by Todd Gamblin, tgamblin@llnl.gov, All rights reserved.
+# LLNL-CODE-647188
+#
+# For details, see https://github.com/llnl/spack
+# Please also see the LICENSE file for our notice and the LGPL.
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU Lesser General Public License (as
+# published by the Free Software Foundation) version 2.1, February 1999.
+#
+# This program is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the IMPLIED WARRANTY OF
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the terms and
+# conditions of the GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with this program; if not, write to the Free Software
+# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+##############################################################################
+from spack import *
+import os
+import shutil
+
+
+class Occa(Package):
+    """OCCA is an open-source (MIT license) library used to program current
+       multi-core/many-core architectures. Devices (such as CPUs, GPUs,
+       Intel's Xeon Phi, FPGAs, etc) are abstracted using an offload-model
+       for application development and programming for the devices is done
+       through a C-based (OKL) or Fortran-based kernel language (OFL).
+       OCCA gives developers the ability to target devices at run-time by
+       using run-time compilation for device kernels.
+    """
+
+    homepage = "http://libocca.org"
+
+    version('develop', git='https://github.com/libocca/occa.git')
+    # FIXME: Replace these hash-versions with tagged version?
+    version('2db622c', git='https://github.com/libocca/occa.git',
+            commit='2db622c480e439b4f60c9491260a1248261b0777')
+    version('0.2', git='https://github.com/libocca/occa.git', tag='0.2')
+    version('0.1', git='https://github.com/libocca/occa.git', tag='0.1')
+    version('0.0.0', git='https://github.com/libocca/occa.git',
+            commmit='381e886886dc87823769c5f20d0ecb29dd117afa')
+
+    variant('cuda',   default=True,
+            description='Activates support for CUDA')
+    variant('openmp',   default=True,
+            description='Activates support for OpenMP')
+    variant('opencl',   default=True,
+            description='Activates support for OpenCL')
+    variant('coi',   default=False,
+            description='Activates support for COI')
+    variant('debug', default=False, description='Enable debug build')
+
+    depends_on('cuda', when='+cuda')
+
+    conflicts('%gcc@6:', when='^cuda@:8')
+    conflicts('%gcc@7:', when='^cuda@:9')
+
+    def install(self, spec, prefix):
+        # The build environment is set by the 'setup_environment' method.
+
+        # Copy the source to the installation directory and build OCCA there.
+        for file in os.listdir('.'):
```

Comment:
Can you use install_tree here?

Reply:
No, `install_tree` uses `shutil.copytree` which does not work when the destination folder exists. It is kind of sad that there is no easy way (or at least I did not find one) to do `cp -rp src/* dest/` directly in python. (Note: the implementation here will copy all files from `src/` whereas `src/*` will not copy files beginning with a dot which a "drawback" of the shell expansion.)